magma/clock.py

```diff
@@ -188,13 +210,14 @@ def get_reset_args(reset_type: Optional[AbstractReset]):
 
 
 def get_default_clocks(defn):
-    clocks = {}
+    default_clocks = {}
     for clock_type in ClockTypes:
-        clocks[clock_type] = first(
-            get_first_clock(port, clock_type)
-            for port in defn.interface.ports.values()
-        )
-    return clocks
+        clocks = get_clocks_from_defn(defn, clock_type)
+        if len(clocks) == 1:
+            default_clocks[clock_type] = clocks[0]
+        else:
+            default_clocks[clock_type] = None
```

Comment:
From above, again, if we use a generator and this `only` method, this code can become
```suggestion
        clocks = get_clocks_from_defn(defn, clock_type)
        try:
            clk = only(clocks)
        except IterableOnlyException:
            clk = None
        default_clocks[clock_type] = clk
```